### PR TITLE
Modernize C++ coding styles

### PR DESCRIPTION
**Pull request by clinssen**
Follow-up to https://github.com/nest/nest-simulator/pull/2493

> 1. Using nullptr instead of NULL or 0 to initialize empty pointers or to check for an undefined pointer.
> 2.  Removing the redundant void argument from function with empty argument lists.
> 3.  Add missing override specifiers for overridden functions in derived classes and, if applicable, remove the redundant virtual specifier when override is set as well.

### CODE REVIEW

File: pynestml/codegeneration/resources_nest/point_neuron/common/SynapseHeader.h.jinja2 [hunk 1]
```diff
@@ -648,7 +648,7 @@ public:
     // use accessor functions (inherited from Connection< >) to obtain delay and target
 {%- if paired_neuron is not none and paired_neuron|length > 0 %}
     {{paired_neuron}}* __target = static_cast<{{paired_neuron}}*>(get_target(tid));
-    assert(__target != NULL);
+    assert(__target != nullptr);
```
Comment (jougs) [inline pynestml/codegeneration/resources_nest/point_neuron/common/SynapseHeader.h.jinja2:651]:
```suggestion
    assert(__target);
```
See #2496

File: pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2 [hunk 2]
```diff
@@ -351,7 +351,7 @@ void {{neuronName}}::init_buffers_()
 {%- endif %}
 {%- if uses_numeric_solver %}
 
-  if ( B_.__s == 0 )
+  if ( B_.__s == nullptr )
```
Comment (JanVogelsang) [inline pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2:354]:
```suggestion
  if ( not B_.__s )
```

Comment (JanVogelsang) [reply pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2:354]: See #2496

File: pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2 [hunk 3]
```diff
@@ -360,7 +360,7 @@ void {{neuronName}}::init_buffers_()
     gsl_odeiv_step_reset( B_.__s );
   }
 
-  if ( B_.__c == 0 )
+  if ( B_.__c == nullptr )
```
Comment (JanVogelsang) [inline pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2:363]:
```suggestion
  if ( not B_.__c )
```
See above

File: pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2 [hunk 4]
```diff
@@ -369,7 +369,7 @@ void {{neuronName}}::init_buffers_()
     gsl_odeiv_control_init( B_.__c, P_.__gsl_error_tol, 0.0, 1.0, 0.0 );
   }
 
-  if ( B_.__e == 0 )
+  if ( B_.__e == nullptr )
```
Comment (JanVogelsang) [inline pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2:372]:
```suggestion
  if ( not B_.__e )
```
See above

File: pynestml/codegeneration/resources_nest/point_neuron/setup/@MODULE_NAME@.h.jinja2 [hunk 5]
```diff
@@ -82,7 +82,7 @@ public:
   /**
    * Return the name of your model.
    */
-  const std::string name( void ) const;
+  const std::string name() const;
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_nest/point_neuron/setup/@MODULE_NAME@.h.jinja2:85]: The `void` keyword must also be removed from the function in the .cpp file.

**Review by JanVogelsang — CHANGES_REQUESTED** [review]
1. There would be some changes needed in the NEST2 templates as well. Should we ignore those as its legacy code anyway?

2. In `NeuronHeader.jinja2` line 741, there is an `||` which should be replaced by an `or`. I couldn't find a way to add a suggestion to that line, probably because it was not part of the changed lines sections.  
3. I could also find 6 occurrences of `&&`, which should be replaced by `and`.

4. In `NeuronHeader.jinja2` line 209, the destructor is missing an `override` specifier.

5. In `SynapseHeader.jinja2`, `handles_test_event` functions are missing `override` specifiers.

File: pynestml/codegeneration/resources_nest/point_neuron/setup/@MODULE_NAME@.h.jinja2 (hunk 5 above)
Comment (JanVogelsang) [reply pynestml/codegeneration/resources_nest/point_neuron/setup/@MODULE_NAME@.h.jinja2:85]: I am confused.. on the _modernize_ branch of @clinssen I can't see the void keyword being present there..

Comment (pnbabu) [reply pynestml/codegeneration/resources_nest/point_neuron/setup/@MODULE_NAME@.h.jinja2:85]: See here: https://github.com/clinssen/nestml/blob/c879b09c7531357904c7dd1bc7a2f63d6a5272b9/pynestml/codegeneration/resources_nest/point_neuron/setup/%40MODULE_NAME%40.cpp.jinja2#L116

Comment (JanVogelsang) [reply pynestml/codegeneration/resources_nest/point_neuron/setup/@MODULE_NAME@.h.jinja2:85]: Ah nevermind, I just didn't read your comment properly. Overlooked the part where you said "in the .cpp file", just read "the void keyword must be removed", looked at the file where you left the comment and was confused. But true, I will create a PR against Charl's branch with the changes I suggested below and also add this suggestion.

**Comment by clinssen** [comment]
Thanks all for the suggestions! @JanVogelsang I merged your PR branch. Could you re-review?

**Review by JanVogelsang — APPROVED** [review]
LGTM!

**Comment by clinssen** [comment]
Many thanks!

**Pull request merged**